xla/hlo/analysis/hlo_replication_analysis_test.cc

```diff
@@ -673,6 +673,69 @@ ENTRY entry {
       FindInstruction(module.get(), "dynamic-slice"), {}, replica_groups_2));
 }
 
+TEST_F(HloReplicationAnalysisTest, PartiallyReplicatedAllGatherFlattenedID) {
+  const std::string module_str = R"(
+HloModule PartiallyReplicatedAllGatherFlattenedID
+
+ENTRY entry {
+  param = s32[2] parameter(0)
+  all-gather0 = s32[4] all-gather(param), dimensions={0}, replica_groups={{0,1},{2,3}}, channel_id=1, use_global_device_ids=true
+  all-gather1 = s32[4] all-gather(param), dimensions={0}, replica_groups={{0,2},{1,3}}, channel_id=2, use_global_device_ids=true
+  ROOT tuple = (s32[4], s32[4]) tuple(all-gather0, all-gather1)
+}
+)";
+
+  TF_ASSERT_OK_AND_ASSIGN(
+      auto module, ParseAndReturnVerifiedModule(module_str, /*replica_count=*/2,
+                                                /*num_partitions=*/2));
+  TF_ASSERT_OK_AND_ASSIGN(
+      std::unique_ptr<HloReplicationAnalysis> replica_analysis,
+      HloReplicationAnalysis::RunWithPartialReplication(
+          module.get(),
+          /*cross_partition_spmd=*/false));
+  TF_ASSERT_OK_AND_ASSIGN(
+      std::unique_ptr<HloReplicationAnalysis> partition_analysis,
+      HloReplicationAnalysis::RunWithPartialReplication(
+          module.get(),
+          /*cross_partition_spmd=*/true));
+
+  std::array<ReplicaGroup, 2> replica_groups0;
+  replica_groups0[0].add_replica_ids(0);
```

Comment:
The `HloInstructionIsReplicatedAt` method is not designed to have duplicated replica IDs. I think the intent of this method is that, assuming cross_partition_spmd=true, the function returns true if and only if for each replica, the instruction output is replicated across the partitions inside the replica. E.g. if there are 2 replicas and 4 partitions and you pass in `replica_groups={{0,2},{1,3}}`, then the function returns true if and only if devices with flattened_id 0 and 2 have the same value, 1 and 3 have the same value, 4 and 6 have the same value, and 5 and 7 have the same value. (Remember that flattened_Id = replia_id * num_partitions + partition_id).

I think it's hard to have a meaningful test with only 2 partitions if cross_partition_spmd=false, and with only 2 replicas if cross_partition_spmd=true.